### ISSUE TITLE
Metricbeat json parser fails against AWS EFS filesystems (or any other over 8 exabytes)

### ISSUE THREAD

**Issue by joey-coleman** [elastic/beats#3687]
Following from the discussion at <https://discuss.elastic.co/t/metricbeat-with-aws-efs/75950>.

AWS EFS filesystems show the volume as having 9223372036854775808 bytes (2^63 bytes) total storage available, which results in warnings like those below in the log when using the `filesystem` and `fsstat` metricbeat modules.

```plain
2017-02-21T19:28:32Z WARN Can not index event (status=400): {"type":"mapper_parsing_exception","reason":"failed to parse [system.filesystem.total]","caused_by":{"type":"json_parse_exception","reason":"Numeric value (9223372036854775808) out of range of long (-9223372036854775808 - 9223372036854775807)\n at [Source: org.elasticsearch.common.bytes.BytesReference$MarkSupportingStreamInputWrapper@6e55ae3; line: 1, column: 398]"}}
2017-02-21T19:28:32Z WARN Can not index event (status=400): {"type":"mapper_parsing_exception","reason":"failed to parse [system.fsstat.total_size.free]","caused_by":{"type":"json_parse_exception","reason":"Numeric value (9223372051157159936) out of range of long (-9223372036854775808 - 9223372036854775807)\n at [Source: org.elasticsearch.common.bytes.BytesReference$MarkSupportingStreamInputWrapper@21e3b3e8; line: 1, column: 269]"}}
```

The relevant subsection of the config is:

```plain
metricbeat.modules:
- module: system
  metricsets:
    - filesystem
    - fsstat
  enabled: true
  period: 30s
```

It is possible to use filtering to exclude events from the `filesystem` metricset, but the `fsstat` produces aggregate values, so will include the EFS mount regardless.

For confirmed bugs, please report:
- Version: `metricbeat version 5.2.1 (amd64), libbeat 5.2.1`
- Operating System: CentOS 7, Linux Kernel 3.10.0
- Steps to Reproduce: 
  1. Instantiate a linux box in AWS
  2. Create an AWS EFS share
  3. Mount the EFS share on the linux box
  4. Start Metricbeat, watch the logs

Output of `df` and `stat -f`, as requested in the original conversation on the fora.

```sh
$ df /mnt/efs
Filesystem                                       1K-blocks     Used        Available Use% Mounted on
fs-xxxxxxxx.efs.us-east-1.amazonaws.com:/ 9007199254740992 19168256 9007199235572736   1% /mnt/efs

$ df -h /mnt/efs
Filesystem                                 Size  Used Avail Use% Mounted on
fs-xxxxxxxx.efs.us-east-1.amazonaws.com:/  8.0E   19G  8.0E   1% /mnt/efs

$ stat -f /mnt/efs
  File: "/mnt/efs"
    ID: 0        Namelen: 255     Type: nfs
Block size: 1048576    Fundamental block size: 1048576
Blocks: Total: 8796093022208 Free: 8796093003489 Available: 8796093003489
Inodes: Total: 0          Free: 0
```

**Comment by jsalcedo09** [comment]
Hello @joey-coleman ,

I had some specific need where I need to meter EFS disk usage and accurately get the value that I'm actually being charged off because of [how AWS EFS meters usage](http://docs.aws.amazon.com/efs/latest/ug/metered-sizes.html).

For this, I have created a very simple beat, [efsbeat](https://github.com/jsalcedo09/efsbeat) (great naming imagination) that actually meters and does the calculation. I have tested with real data and it seems to get the exact value EFS meters, as well as what the systems see (both are different values).

I know this is not a bugfix type of comment but might help you achieve your goal

Thanks!

**Comment by joey-coleman** [comment]
Thanks @jsalcedo09, that's a useful pointer — I don't actually need reporting on the EFS mounts I have at all right now, and only reported the bug because it is a bug and filtering out the EFS mount is a bit kludgy.  (I need reporting from the `filesystem` module for other filesystems, just not the EFS mounts.)

**Comment by ruflin** [comment]
@jsalcedo09 I must have a look at your beat. How did you handle the extra large values?

**Comment by jsalcedo09** [comment]
Hello @ruflin ,

I'm actually not calculating any kind of metric, like percentage usage, where I would need to know the total size (where this thread is all about). I'm just doing some folder size calculation, but including AWS EFS metering rules.

All I'm doing [is walking the directory and calculating size](https://github.com/jsalcedo09/efsbeat/blob/master/beater/efsbeat.go#L92) pretty simple and we might get not the best performance:

```go
//EfsRoundValue AWS EFS size in bytes that files and folders will be rounded to 4Kb
const EfsRoundValue int64 = 4 * 1024
//EfsMetadata AWS EFS Size in bytes of the files metadata 2Kb
const EfsMetadata int64 = 2 * 1024

...

var realSize int64
var efsSize int64
logp.Info("Calculating path %s size...", path)
err := filepath.Walk(path, func(_ string, info os.FileInfo, err error) error {
	var s = info.Size()
	realSize += s
	if s <= 0 {
                //This is done to keep s > 0 so calculation bellow rounds up to the next 4Kb increment
		s = EfsMetadata
	}
	if info.IsDir() {
		efsSize += ((int64(math.Ceil(float64(s) / float64(EfsRoundValue)))) * EfsRoundValue)
	} else {
		efsSize += ((int64(math.Ceil(float64(s) / float64(EfsRoundValue)))) * EfsRoundValue) + EfsMetadata
	}
	return err
})
```

As far as I can see, there is a problem with Elasticsearch itself, where the biggest value type it can store is a `long` signed value, where they can go from `-2^63 to 2^63 - 1`, unfortunately, this is a one byte issue!!.

Hope this help, but I know we are talking about two different things.

Thanks!

**Comment by Nodens2k** [comment]
Just for the record, if you connect to a AWS Storage Gateway server via NFS, and the filesystem behind is a large enough S3 instance, the mount information reports 2^63 available bytes, too.

```
192.168.1.49:/env-prod       9007199254740992        0 9007199254740992   0% /mnt/env-prod
192.168.1.49:/env-dev        9007199254740992        0 9007199254740992   0% /mnt/env-dev
192.168.1.49:/env-stage      9007199254740992        0 9007199254740992   0% /mnt/env-stage
```

**Comment by ycombinator** [comment]
A similar issue was reported by this user on discuss: https://discuss.elastic.co/t/numeric-value-18446744073709551615-out-of-range-of-long-9223372036854775808-9223372036854775807/155732

**Comment by alvarolobato** [comment]
This problem should now be fixed after (https://github.com/elastic/beats/pull/8991) which will be included in 7.0
